subsys/net/lib/coap/coap_client.c

```diff
@@ -0,0 +1,716 @@
+{
+	int ret = 0;
+	int response_type;
+	int block_option;
+	int block_num;
+	bool blockwise_transfer = false;
+	bool last_block = false;
+
+	/* Handle different types, ACK might be separate or piggybacked
+	 * CON and NCON contains a separate response, CON needs an empty response
+	 * CON request results as ACK and possibly separate CON or NCON response
+	 * NCON request results only as a separate CON or NCON message as there is no ACK
+	 * With RESET, just drop gloves and call the callback.
+	 */
+	response_type = coap_header_get_type(response);
+
+	/* Reset and Ack need to match the message ID with request */
+	if ((response_type == COAP_TYPE_ACK || response_type == COAP_TYPE_RESET) &&
+	     coap_header_get_id(response) != client->pending.id)  {
+		LOG_ERR("Unexpected ACK or Reset");
+		return -EFAULT;
+	} else if (response_type == COAP_TYPE_RESET) {
+		coap_pending_clear(&client->pending);
+	}
+
+	/* CON, NON_CON and piggybacked ACK need to match the token with original request */
+	uint16_t payload_len;
+	uint8_t response_code = coap_header_get_code(response);
+	const uint8_t *payload = coap_packet_get_payload(response, &payload_len);
+
+	/* Separate response */
+	if (payload_len == 0 && response_type == COAP_TYPE_ACK &&
+	    response_code == COAP_CODE_EMPTY) {
+		/* Clear the pending, poll uses now the separate timeout for the response. */
+		coap_pending_clear(&client->pending);
+		return 1;
+	}
+
+	/* Check for tokens */
+	if (!token_compare(client, response)) {
+		LOG_ERR("Not matching tokens, respond with reset");
+		ret = send_reset(client, response, COAP_RESPONSE_CODE_NOT_FOUND);
+		return 1;
+	}
+
+	/* Send ack for CON */
+	if (response_type == COAP_TYPE_CON) {
+		/* CON response is always a separate response, respond with empty ACK. */
+		ret = send_ack(client, response, COAP_CODE_EMPTY);
+		if (ret < 0) {
+			goto fail;
+		}
+	}
+
+	if (client->pending.timeout != 0) {
+		coap_pending_clear(&client->pending);
+	}
+
+	/* Check if block2 exists */
+	block_option = coap_get_option_int(response, COAP_OPTION_BLOCK2);
+	if (block_option > 0) {
+		blockwise_transfer = true;
+		last_block = !GET_MORE(block_option);
+		block_num = GET_BLOCK_NUM(block_option);
+
+		if (block_num == 0) {
+			coap_block_transfer_init(&client->recv_blk_ctx, COAP_BLOCK_128, 0);
+			client->offset = 0;
+		}
+
+		ret = coap_update_from_block(response, &client->recv_blk_ctx);
+		if (ret < 0) {
+			LOG_ERR("Error updating block context");
+		}
+		coap_next_block(response, &client->recv_blk_ctx);
+	} else {
+		client->offset = 0;
+		last_block = true;
+	}
+
+	/* Check if this was a response to last blockwise send */
+	if (client->send_blk_ctx.total_size > 0) {
+		blockwise_transfer = true;
+		if (client->send_blk_ctx.total_size == client->send_blk_ctx.current) {
+			last_block = true;
+		} else {
+			last_block = false;
+		}
+	}
+
+	/* Call user callback */
+	if (client->coap_request->cb) {
+		uint8_t result_code = coap_header_get_code(response);
```

Comment:
`result_code` is a duplicate of `response_code` -- they are both set to `coap_header_get_code(response)`.